### PR TITLE
`IndicesInfo` for succinct gathering of info on indices.

### PR DESCRIPTION
**Pull request by Tokazama**
This tidies up the index traits into a single call, `IndicesInfo` and helps avoid directly indexing the `parameters` field within the generated method of `to_indices` (since access to that field directly is discouraged).

### CODE REVIEW

File: lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl
Comment (chriselrod) [inline lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl:580]: Seems unfortunate that this sort of thing has to be a generated function.
If we have a lot of code that wants to do something like this, it may be worth defining a `typetail` to take a recursive approach.

File: src/indexing.jl
Comment (chriselrod) [inline src/indexing.jl:141]: comments could use some rephrasing

**Review by chriselrod — APPROVED** [review]

File: lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl
Comment (Tokazama) [reply lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl:580]: The generated function itself isn't costing very much in load time but if we end up needing to do the same thing a lot we could have a single generated `each_field(f, ::Type{<:Tuple})` to avoid defining too many of these. What we need is essentially `ntuple(i -> f(fieldtype(T, i)), Val{fieldcount(T)}())` but I've noticed constant propagation gives up pretty easily with `fieldtype`.

Either way, I should at least try being proactive about this before merging so I'll try a few things out before merging this.

Comment (chriselrod) [reply lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl:580]: While I do think latency is a major issue, I am more worried about const prop failures.
(And many latency problems do come from type instabilities.)

Comment (Tokazama) [reply lib/ArrayInterfaceCore/src/ArrayInterfaceCore.jl:580]: I tried using this code instead:
```julia
struct IndexInfo{NI,NS,IS} end
IndexInfo(x) = IndexInfo{ndims_index(x),ndims_shape(x),is_splat_index(x)}()

indices_info(@nospecialize x::Tuple) = indices_info(typeof(x))
function indices_info(::Type{T}) where {T<:Tuple}
    ntuple(i -> IndexInfo(fieldtype(T, i)), Val{fieldcount(T)}())
end
```

and inference was fine. I'm not completely confident it will hold up in practice because at the very least it will be used for `ArrayInterface._to_indices` which is already several steps away from the call site `ArrayInterface.getindex`.

If there's something else I'm missing I'm open to trying it out.

**Pull request merged**
